### PR TITLE
feat(api-v3): update conversations access roles FS-1269

### PR DESCRIPTION
**Pull request by David-Henner**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issue

In **v2** and earlier versions `access_role` represents the legacy access role value. While `access_role_v2` represents the "new" access role values.

In **v3** the field `access_role_v2` has been removed from the payload of `Conversation` objects.
Instead, the field `access_role` is used. **v3** doesn't support legacy access roles anymore, hence why only one `access_role` field is used.

### Solution

- Created a new protocol `CodableAPIVersionAware`, to implement custom encoding & decoding based on the api version.
- Conformed `Payload.NewConversation` and `Payload.Conversation` to this protocol and encoded / decoded proper access roles fields based on api version
- Also covered the special case of decoding conversations from update events, since their format is not version dependent.
- Updated `ConversationRequestStrategy` to use api version aware coding

### Testing

- [x] Added tests

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Review by agisilaos — DISMISSED** [review]

File: wire-ios-request-strategy/Sources/Payloads/Payload+CodingTests.swift [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    var mockEncode: MockEncode?
+
+    func encode(to encoder: Encoder) throws {
+        mockEncode?(encoder)
+    }
+
+    init() { }
+}
+
+class Payload_CodingTests: XCTestCase {
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Payloads/Payload+CodingTests.swift:66]: why is this file inside Sources, maybe because of monorepo architecture?

**Review by netbe — DISMISSED** [review]
LGTM;) I was suprised to see some tests under Sources but I guess that's the monorepo

File: wire-ios-request-strategy/Sources/Payloads/Payload+CodingTests.swift (hunk 1 above)
Comment (netbe) [reply wire-ios-request-strategy/Sources/Payloads/Payload+CodingTests.swift:66]: should be under wire-ios-request-strategy/Tests/Sources/ no?

File: wire-ios-request-strategy/Sources/Payloads/Payload+ConversationTests.swift [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+@testable import WireRequestStrategy
+import XCTest
+
+class Payload_ConversationTests: MessagingTestBase {
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Payloads/Payload+ConversationTests.swift:22]: same move to wire-ios-request-strategy/Tests)/Sources/

**Comment by johnxnguyen** [comment]
> LGTM;) I was suprised to see some tests under Sources but I guess that's the monorepo

@netbe i think that's just the request strategy project, for some reason tests are next to source files.

**Comment by David-Henner** [comment]
> > LGTM;) I was suprised to see some tests under Sources but I guess that's the monorepo
> 
> @netbe i think that's just the request strategy project, for some reason tests are next to source files.

Yep, not sure why that is but I figured I'd do the same for consistency

**Comment by David-Henner** [comment]
Added the changes from https://github.com/wireapp/wire-ios-sync-engine/pull/1820

**Review by agisilaos — DISMISSED** [review]

**Review by netbe — DISMISSED** [review]

File: wire-ios-request-strategy/Sources/Payloads/Processing/PayloadProcessing+Conversation.swift [hunk 3]
```diff
@@ -228,13 +228,17 @@ extension Payload.Conversation {
...
 
         if let accessModes = access {
-            if let accessRoles = accessRoleV2 {
+            if let accessRoles = accessRoles {
                 conversation.updateAccessStatus(accessModes: accessModes, accessRoles: accessRoles)
-            } else if let accessRole = accessRole, let legacyAccessRole = ConversationAccessRole(rawValue: accessRole) {
+            } else if
+                let accessRole = legacyAccessRole,
+                let legacyAccessRole = ConversationAccessRole(rawValue: accessRole)
+            {
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Payloads/Processing/PayloadProcessing+Conversation.swift]:
```suggestion
            } else if let accessRole = legacyAccessRole,
            let legacyAccessRole = ConversationAccessRole(rawValue: accessRole)
            {
```

**Review by netbe — DISMISSED** [review]
LGTM;) just formatting of else if

**Review by netbe — APPROVED** [review]

**Review by agisilaos — APPROVED** [review]

**Pull request merged**